File: fibonacci_test.c
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include "fibonacci.c"

int main() {
    int test_values[] = {0, 1, 5, 10, 20, 30, 40};
    int num_tests = sizeof(test_values) / sizeof(test_values[0]);

    printf("Testing Fibonacci methods:\n\n");

    for (int i = 0; i < num_tests; i++) {
        int n = test_values[i];
        uint64_t ops = 0;

        printf("n = %d\n", n);

        // Iterative
        ops = 0;
        uint64_t iter_result = fib_iterative(n, &ops);
        printf("  Iterative: %" PRIu64 " (Ops: %" PRIu64 ")\n", iter_result, ops);

        // Dynamic Programming
        ops = 0;
        uint64_t dp_result = fib_dynamic(n, &ops);
        printf("  Dynamic:   %" PRIu64 " (Ops: %" PRIu64 ")\n", dp_result, ops);

        // Recursive (only for small n)
        if (n <= 30) {
            ops = 0;
            uint64_t rec_result = fib_recursive(n, &ops);
            printf("  Recursive: %" PRIu64 " (Ops: %" PRIu64 ")\n", rec_result, ops);
        } else {
            printf("  Recursive: Skipped (n > 30)\n");
        }

        printf("\n");
    }

    return 0;
}

File: fibonacci_test.py
from fibonacci import fib_iterative, fib_dynamic, fib_recursive

def test_fibonacci_methods():
    test_values = [0, 1, 2, 5, 10, 20, 30]
    print("Running tests for Fibonacci methods...\n")
    
    # Test each method and compare results
    for n in test_values:
        iter_result, _ = fib_iterative(n)
        dyn_result, _ = fib_dynamic(n)
        rec_result = fib_recursive(n, [0]) if n <= 30 else iter_result  # Skip slow recursion

        print(f"n = {n}")
        print(f"  Iterative: {iter_result}")
        print(f"  Dynamic:   {dyn_result}")
        if n <= 30:
            print(f"  Recursive: {rec_result}")
        else:
            print("  Recursive: Skipped (n > 30)")

        if iter_result == dyn_result == rec_result:
            print("  ✅ All methods match\n")
        else:
            print("  ❌ Mismatch detected!\n")

if __name__ == "__main__":
    test_fibonacci_methods()
